Boundary Docs: Configure workers refactor (#3077)
hashicorp/web-unified-docs: content/boundary/v1.0.x/content/docs/workers/registration/kms.mdx

---
layout: docs
page_title: Register a worker using an external KMS
description: >-
  Register a Boundary worker using an external KMS. Workers that share a KMS key with their upstream register automatically, which supports automatic scaling.
# START AUTO GENERATED METADATA, DO NOT EDIT
created_at: 2026-08-10T22:18:18.000Z
last_modified: 2026-08-10T22:18:18.000Z
# END AUTO GENERATED METADATA
---

# Register a worker using an external KMS

In the external KMS method, the worker authenticates to its upstream using a key management system (KMS) key that both share.
This method registers the worker in addition to authenticating it, so it requires no API call.

Use this method for automatic scaling.

> Workers that use an external KMS re-authenticate each time they connect, so they do not store credentials on disk and you can replace them freely.

Reviewer: Same comment here on KMS creds/ auth

Author: I updated to this, do you think it's accurate now?

```
Workers that use an external KMS store their credentials in memory and generate new ones on restart.
```

</Note>

## Add the worker-auth KMS block

Workers that register through an external KMS require a `kms` block with the `worker-auth` purpose.
The upstream controller or worker must have a `kms` block that references the same key and purpose.

The following example shows a complete worker configuration:

<CodeBlockConfig filename="/etc/boundary.d/egress-worker.hcl">

```hcl
disable_mlock = true

listener "tcp" {
  address = "0.0.0.0:9202"
  purpose = "proxy"
}

worker {
  name              = "kms-worker-1"
  description       = "Worker registered with an external KMS"
  public_addr       = "10.0.0.10:9202"
  initial_upstreams = ["10.0.0.1:9201"]

  tags {
    type = ["prod", "kms"]
  }
}

# This key must match the worker-auth key on the upstream controller
kms "awskms" {
  purpose    = "worker-auth"
  region     = "us-east-1"
  kms_key_id = "19ec80b0-dfdd-4d97-8164-c6examplekey1"
}
```

</CodeBlockConfig>

If a controller and worker run as the same server process, you only need one `kms` block with the `worker-auth` purpose.

> Workers that register through an external KMS do not require an `auth_storage_path`, because the KMS re-authenticates them each time they connect.

Reviewer: ```suggestion
Workers that register through an external KMS do not require an `auth_storage_path`, because the KMS stores credentials in memory.
```
